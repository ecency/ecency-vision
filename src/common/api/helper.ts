import defaults from "../constants/defaults.json";

export const apiBase = (endpoint: string): string => `${defaults.base}${endpoint}`;

export const apiBaseImage = (endpoint: string): string => `${defaults.imageServer}${endpoint}`;
