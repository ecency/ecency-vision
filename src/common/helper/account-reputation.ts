const isHumanReadable = (input: number): boolean => {
  return Math.abs(input) > 0 && Math.abs(input) <= 100;
};

export default (input: string | number): number => {
  if (typeof input === 'number' && isHumanReadable(input)) {
    return Math.floor(input);
  }

  if (typeof input === 'string') {
    input = Number(input);

    if (isHumanReadable(input)) {
      return Math.floor(input);
    }
  }

  if (input === 0) {
    return 25;
  }

  let neg = false;

  if (input < 0) neg = true;

  let reputationLevel = Math.log10(Math.abs(input));
  reputationLevel = Math.max(reputationLevel - 9, 0);

  if (reputationLevel < 0) reputationLevel = 0;

  if (neg) reputationLevel *= -1;

  reputationLevel = reputationLevel * 9 + 25;

  return Math.floor(reputationLevel);
};
