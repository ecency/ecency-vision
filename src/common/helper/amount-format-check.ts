export default (v: string) => /^\d{1,3}(,?\d{3})*(\.(\d{3},?)*\d{1,3})?$/.test(v);
