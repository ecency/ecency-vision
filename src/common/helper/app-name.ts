import stripTags from "../util/strip-tags";

interface ObjInput {
  name?: string;
}

export default (input: string | null | undefined | ObjInput) => {
  if (!input) {
    return "";
  }

  if (typeof input === "string") {
    return stripTags(input);
  }

  if (typeof input === "object" && input.name !== undefined) {
    return stripTags(input.name);
  }

  return "";
};
