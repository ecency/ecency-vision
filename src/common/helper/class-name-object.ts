export function classNameObject(obj: Record<string, boolean | undefined>): string {
  return Object.entries(obj)
    .filter(([className, isActive]) => !!isActive)
    .map(([className]) => className)
    .join(" ");
}
