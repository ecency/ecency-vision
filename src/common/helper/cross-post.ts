import {postBodySummary} from '@ecency/render-helper';
import {Entry} from '../store/entries/types';

export const crossPostMessage = (body: string) => {
  const crossPostRegex =
    /^This is a cross post of \[@(.*?)\/(.*?)\]\(\/.*?@.*?\/.*?\) by @.*?\.<br>/;

  if (body.match(crossPostRegex)) {
    const message = body.replace(crossPostRegex, '');
    return postBodySummary(message);
  }

  return null;
};

export const makeCrossPostMessage = (
  entry: Entry,
  poster: string,
  message: string,
) => {
  return `This is a cross post of [@${entry.author}/${entry.permlink}](/${entry.category}/@${entry.author}/${entry.permlink}) by @${poster}.<br><br>${message}`;
};
