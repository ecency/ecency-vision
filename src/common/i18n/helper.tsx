import React from "react";

import xss from "xss";

import { _t } from "./index";

export const HTML_WHITELIST = {
  a: ["class", "href", "target", "rel"],
  span: ["class"],
  h3: [],
  p: [],
  strong: [],
  b: [],
  img: ["src", "alt"],
  ul: [],
  li: [],
  br: []
};

export const safeHtml = (html: string): string => {
  return xss(html, {
    whiteList: HTML_WHITELIST,
    stripIgnoreTagBody: true,
    stripIgnoreTag: true,
    css: false
  });
};

interface Props {
  k: string;
  args?: {};
  children: JSX.Element;
}

export class Tsx extends React.Component<Props> {
  render() {
    const { k, args } = this.props;

    const { children } = this.props;

    const html = safeHtml(_t(k, args));

    return React.cloneElement(children, { dangerouslySetInnerHTML: { __html: html } });
  }
}
