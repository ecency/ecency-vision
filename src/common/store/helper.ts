import {Store} from 'redux';

import i18n from 'i18next';

import {AppState} from './index';
import {ActiveUser, UserPoints} from './active-user/types';
import {
  loginAct as loginActiveUser,
  logoutAct as logoutActiveUser,
  updateAct as updateActiveUserAct,
} from './active-user';

import {getAccount, getDynamicProps} from '../api/hive';
import {getPoints, usrActivity, getPromotedEntries} from '../api/private-api';
import {reloadAct as reloadUsers} from './users';
import {fetchedAct as loadDynamicProps} from './dynamic-props';
import {fetchedAct as entriesFetchedAct} from './entries';
import {
  setCurrencyAct as setCurrency,
  muteNotificationsAct as muteNotifications,
  setLangAct as setLang,
  setNsfwAct as setNsfw,
} from './global';

import {getCurrencyRate} from '../api/misc';

import currencies from '../constants/currencies.json';

import * as ls from '../../common/util/local-storage';

import currencySymbol from '../../common/helper/currency-symbol';

import {AppWindow} from '../../client/window';

declare var window: AppWindow;

export const activeUserMaker = (
  name: string,
  points: string = '0.000',
  uPoints: string = '0.000',
): ActiveUser => {
  return {
    username: name,
    data: {name},
    points: {
      points,
      uPoints,
    },
  };
};

export const activeUserUpdater = async (store: Store<AppState>) => {
  const state = store.getState();
  if (state.activeUser) {
    const {username} = state.activeUser;

    let account;
    try {
      account = await getAccount(username);
    } catch (e) {
      return;
    }

    let points: UserPoints;
    try {
      const p = await getPoints(username);
      points = {points: p.points, uPoints: p.unclaimed_points};
    } catch (e) {
      points = {
        points: '0.000',
        uPoints: '0.000',
      };
    }

    store.dispatch(updateActiveUserAct(account, points));
  }
};

export const syncActiveUser = (store: Store<AppState>) => {
  const state = store.getState();

  const activeUser = ls.get('active_user');

  // logout
  if (!activeUser && state.activeUser) {
    store.dispatch(logoutActiveUser());
    return;
  }

  // login
  if (activeUser && !state.activeUser) {
    store.dispatch(loginActiveUser());
    activeUserUpdater(store).then();
    return;
  }

  // switch
  if (
    activeUser &&
    state.activeUser &&
    activeUser !== state.activeUser.username
  ) {
    store.dispatch(loginActiveUser());
    activeUserUpdater(store).then();
    return;
  }
};

export const clientStoreTasks = (store: Store<AppState>) => {
  // To use in places where we can't access to store
  window.usePrivate = store.getState().global.usePrivate;

  // Initial state from browser's local storage
  store.dispatch(reloadUsers());
  store.dispatch(loginActiveUser());

  // Load dynamic props
  getDynamicProps().then(resp => {
    store.dispatch(loadDynamicProps(resp));
  });

  // Update active user in interval
  activeUserUpdater(store).then();
  setInterval(() => {
    activeUserUpdater(store).then();
  }, 60 * 1000);

  // Active user sync between tabs / windows
  setInterval(() => {
    syncActiveUser(store);
  }, 2000);

  // Do check-in in interval
  const checkIn = () => {
    const state = store.getState();
    if (state.activeUser) {
      usrActivity(state.activeUser?.username!, 10).then();
    }
  };

  // wait for initial user sync
  setTimeout(() => {
    checkIn();
  }, 5000);
  setInterval(checkIn, 1000 * 60 * 15 + 8);

  // Inject / update promoted entries to store
  const promotedEntries = () => {
    getPromotedEntries().then(r => {
      store.dispatch(entriesFetchedAct('__promoted__', r, false));
    });
  };
  promotedEntries();
  setInterval(promotedEntries, 1000 * 60 * 5);

  // Currency
  const currency = ls.get('currency');
  if (currency && currencies.find(x => x.id === currency)) {
    const symbol = currencySymbol(currency);
    getCurrencyRate(currency).then(rate => {
      store.dispatch(setCurrency(currency, rate, symbol));
    });
  }

  // Notifications
  if (ls.get('notifications') === false) {
    store.dispatch(muteNotifications());
  }

  // Language
  const lang = ls.get('lang');
  if (lang) {
    if (store.getState().global.lang !== lang) {
      i18n.changeLanguage(lang).then(() => {
        store.dispatch(setLang(lang));
      });
    }
  }

  // NSFW
  if (ls.get('nsfw') === true) {
    store.dispatch(setNsfw(true));
  }
};
