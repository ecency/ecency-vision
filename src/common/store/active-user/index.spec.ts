import reducer, { initialState, loginAct, logoutAct, updateAct } from './index';

import { fullAccountInstance } from '../../helper/test-helper';

jest.mock('../../util/local-storage', () => ({
  get: (k: string) => {
    if (k === 'active_user') {
      return 'foo';
    }

    if (k.startsWith('user_')) {
      return '....';
    }

    return null;
  }
}));

let state = initialState;

it('1- default state', () => {
  expect(state).toMatchSnapshot();
});

it('2- loginAct', () => {
  state = reducer(state, loginAct());
  expect(state).toMatchSnapshot();
});

it('3- updateAct', () => {
  state = reducer(
    state,
    updateAct({ ...fullAccountInstance, name: 'foo' }, { points: '0.100', uPoints: '0.200' })
  );
  expect(state).toMatchSnapshot();
});
