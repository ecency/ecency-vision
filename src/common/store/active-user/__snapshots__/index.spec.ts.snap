// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`1- default state 1`] = `null`;

exports[`2- loginAct 1`] = `
Object {
  "data": Object {
    "name": "foo",
  },
  "points": Object {
    "points": "0.000",
    "uPoints": "0.000",
  },
  "username": "foo",
}
`;

exports[`3- updateAct 1`] = `
Object {
  "data": Object {
    "__loaded": true,
    "active": Object {
      "account_auths": Array [],
      "key_auths": Array [
        Array [
          "STM71z4rmzGHdp7pmePZyS1G2GrbuBqg9PPHRgiK6uYqjzUTRNCRR",
          1,
        ],
      ],
      "weight_threshold": 1,
    },
    "balance": "1.751 HIVE",
    "created": "2018-01-29T20:10:06",
    "delegated_vesting_shares": "145395.758709 VESTS",
    "downvote_manabar": Object {
      "current_mana": "36303329010",
      "last_update_time": 1609750587,
    },
    "follow_stats": Object {
      "account": "talhasch",
      "follower_count": 232,
      "following_count": 27,
    },
    "hbd_balance": "0.311 HBD",
    "json_metadata": "{\\"profile\\":{\\"name\\":\\"Talha B.\\",\\"about\\":\\"Founder @runkod, Software Developer @esteemapp, Lifelong Learner\\",\\"cover_image\\":\\"https://img.esteem.app/rwd380.jpg\\",\\"profile_image\\":\\"https://img.esteem.app/821e0q.jpg\\",\\"website\\":\\"https://github.com/talhasch\\",\\"location\\":\\"Istanbul\\"},\\"escrow\\":{\\"fees\\":{\\"STEEM\\":\\"1.000\\",\\"SBD\\":\\"3.000\\"},\\"terms\\":\\"lorem ipsum dolor sit amet\\"}}",
    "last_post": "2020-06-01T05:53:33",
    "last_vote_time": "2020-06-01T22:22:22",
    "memo_key": "STM7rioGL7NopT2Zo446hnzAiHLp5sJ1gS55xu3NikmxfQnEznsue",
    "name": "foo",
    "next_vesting_withdrawal": "1969-12-31T23:59:59",
    "owner": Object {
      "account_auths": Array [],
      "key_auths": Array [
        Array [
          "STM7F7zfd6ieangxz6uxQkYifUS5H841x5E41SYZaVc9F9cGPJ9jN",
          1,
        ],
      ],
      "weight_threshold": 1,
    },
    "pending_claimed_accounts": 0,
    "post_count": 261,
    "posting": Object {
      "account_auths": Array [
        Array [
          "ecency.app",
          1,
        ],
        Array [
          "esteem-app",
          1,
        ],
        Array [
          "esteem.app",
          1,
        ],
        Array [
          "esteemapp",
          1,
        ],
        Array [
          "peakd.app",
          1,
        ],
      ],
      "key_auths": Array [
        Array [
          "STM6uvU7j624wCZNa2pcXvkqmbnNC1cgraDrVG3pFRhXvj9LYS7Xp",
          1,
        ],
      ],
      "weight_threshold": 1,
    },
    "posting_json_metadata": "{\\"profile\\":{\\"name\\":\\"Talha B.\\",\\"about\\":\\"Founder @runkod, Software Developer @ecency, Lifelong Learner\\",\\"cover_image\\":\\"https://images.ecency.com/DQmTsFJJV7gXpsHNatEWz1vgzykaM4WByxAgLMzWshmXNrj/cover.jpg\\",\\"profile_image\\":\\"https://images.ecency.com/DQmV2J1oAUoKxjyJQTu3TcZfGxDr5fzajMGhHEd1e1GALca/ben.jpg\\",\\"website\\":\\"https://github.com/talhasch\\",\\"location\\":\\"Istanbul\\",\\"version\\":2}}",
    "profile": Object {
      "about": "Founder @runkod, Software Developer @ecency, Lifelong Learner",
      "cover_image": "https://images.ecency.com/DQmTsFJJV7gXpsHNatEWz1vgzykaM4WByxAgLMzWshmXNrj/cover.jpg",
      "location": "Istanbul",
      "name": "Talha B.",
      "profile_image": "https://images.ecency.com/DQmV2J1oAUoKxjyJQTu3TcZfGxDr5fzajMGhHEd1e1GALca/ben.jpg",
      "website": "https://github.com/talhasch",
    },
    "proxied_vsf_votes": Array [
      0,
      0,
      0,
      0,
    ],
    "proxy": "",
    "received_vesting_shares": "0.000000 VESTS",
    "recovery_account": "",
    "reputation": "1021265023022",
    "reward_hbd_balance": "0.000 HBD",
    "reward_hive_balance": "0.000 HIVE",
    "reward_vesting_balance": "0.000000 VESTS",
    "reward_vesting_hive": "0.000 HIVE",
    "savings_balance": "1.000 HIVE",
    "savings_hbd_balance": "0.002 HBD",
    "savings_hbd_last_interest_payment": "2021-03-18T15:42:03",
    "savings_hbd_seconds": "1235678",
    "savings_hbd_seconds_last_update": "2021-03-18T15:42:03",
    "to_withdraw": "87017000000",
    "vesting_shares": "151590.952150 VESTS",
    "vesting_withdraw_rate": "0.000000 VESTS",
    "voting_manabar": Object {
      "current_mana": "6195193441",
      "last_update_time": 1607690265,
    },
    "voting_power": 9476,
    "withdrawn": "87017000000",
    "witness_votes": Array [
      "blocktrades",
      "good-karma",
      "gtg",
    ],
  },
  "points": Object {
    "points": "0.100",
    "uPoints": "0.200",
  },
  "username": "foo",
}
`;
